bytecomp/translmod.ml

```diff
@@ -88,22 +88,27 @@ let rec apply_coercion loc strict restr arg =
   match restr with
     Tcoerce_none ->
       arg
-  | Tcoerce_structure(pos_cc_list, id_pos_list) ->
+  | Tcoerce_structure(pos_cc_list, id_pos_list, runtime_fields) ->
+      assert (List.length runtime_fields = List.length pos_cc_list);
+      let names = Array.of_list runtime_fields in
       name_lambda strict arg (fun id ->
-        let get_field pos = Lprim(Pfield (pos, Fld_na (*TODO*)),[Lvar id], loc) in
+        let get_field_i i pos =
+          Lprim(Pfield (pos, Fld_module names.(i)),[Lvar id], loc) in
+        let get_field_name name pos =
+          Lprim(Pfield (pos, Fld_module name),[Lvar id], loc) in
         let lam =
-          Lprim(Pmakeblock(0, Lambda.default_tag_info, Immutable),
-                List.map (apply_coercion_field loc get_field) pos_cc_list, loc)
+          Lprim(Pmakeblock(0, Blk_module runtime_fields, Immutable),
+                List.mapi (fun i x -> (apply_coercion_field loc (get_field_i i) x)) pos_cc_list, loc)
         in
-        wrap_id_pos_list loc id_pos_list get_field lam)
+        wrap_id_pos_list loc id_pos_list get_field_name lam)
```

Comment:
This may be a good place to some simple optimizations, check Belt_Id.MakeComparableU, but let's save it for next iteration